### ISSUE TITLE
Flaky `TestDistributor/caching_unmarshal_data_enabled/series_with_exemplars`

### ISSUE THREAD

**Issue by dimitarvdimitrov** [grafana/mimir#7448]
The test failed on a PR which doesn't work with exemplars or the distributor

https://github.com/grafana/mimir/actions/runs/8016409523/job/21899113333?pr=7173

```
=== RUN   TestDistributor/caching_unmarshal_data_enabled/series_with_exemplars
08:44:30 querier: ts=2024-02-23T08:44:30.956257016Z caller=loader.go:116 level=warn msg="bucket index not found" user=e2e-user
    distributor_test.go:321: 
        	Error Trace:	/home/runner/work/mimir/mimir/integration/distributor_test.go:321
        	Error:      	Not equal: 
        	            	expected: []v1.ExemplarQueryResult{v1.ExemplarQueryResult{SeriesLabels:model.LabelSet{"__name__":"foobar_with_exemplars"}, Exemplars:[]v1.Exemplar{v1.Exemplar{Labels:model.LabelSet{"test":"test"}, Value:[123](https://github.com/grafana/mimir/actions/runs/8016409523/job/21899113333?pr=7173#step:15:124), Timestamp:1708674267000}}}}
        	            	actual  : []v1.ExemplarQueryResult{}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,16 +1,2 @@
        	            	-([]v1.ExemplarQueryResult) (len=1) {
        	            	- (v1.ExemplarQueryResult) {
        	            	-  SeriesLabels: (model.LabelSet) (len=1) {
        	            	-   (model.LabelName) (len=8) "__name__": (model.LabelValue) (len=21) "foobar_with_exemplars"
        	            	-  },
        	            	-  Exemplars: ([]v1.Exemplar) (len=1) {
        	            	-   (v1.Exemplar) {
        	            	-    Labels: (model.LabelSet) (len=1) {
        	            	-     (model.LabelName) (len=4) "test": (model.LabelValue) (len=4) "test"
        	            	-    },
        	            	-    Value: (model.SampleValue) 123,
        	            	-    Timestamp: (model.Time) 1708674267000
        	            	-   }
        	            	-  }
        	            	- }
        	            	+([]v1.ExemplarQueryResult) {
        	            	 }
        	Test:       	TestDistributor/caching_unmarshal_data_enabled/series_with_exemplars
```

**Comment by dimitarvdimitrov** [comment]
there's another instance of the same test a few minutes apart https://github.com/grafana/mimir/actions/runs/8017427753/job/21902224686?pr=7446

```
=== RUN   TestDistributor/caching_unmarshal_data_disabled/series_with_exemplars
    distributor_test.go:321: 
        	Error Trace:	/home/runner/work/mimir/mimir/integration/distributor_test.go:321
        	Error:      	Not equal: 
        	            	expected: []v1.ExemplarQueryResult{v1.ExemplarQueryResult{SeriesLabels:model.LabelSet{"__name__":"foobar_with_exemplars"}, Exemplars:[]v1.Exemplar{v1.Exemplar{Labels:model.LabelSet{"test":"test"}, Value:123, Timestamp:1708679924000}}}}
        	            	actual  : []v1.ExemplarQueryResult{}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,16 +1,2 @@
        	            	-([]v1.ExemplarQueryResult) (len=1) {
        	            	- (v1.ExemplarQueryResult) {
        	            	-  SeriesLabels: (model.LabelSet) (len=1) {
        	            	-   (model.LabelName) (len=8) "__name__": (model.LabelValue) (len=21) "foobar_with_exemplars"
        	            	-  },
        	            	-  Exemplars: ([]v1.Exemplar) (len=1) {
        	            	-   (v1.Exemplar) {
        	            	-    Labels: (model.LabelSet) (len=1) {
        	            	-     (model.LabelName) (len=4) "test": (model.LabelValue) (len=4) "test"
        	            	-    },
        	            	-    Value: (model.SampleValue) 123,
        	            	-    Timestamp: (model.Time) 1708679924000
        	            	-   }
        	            	-  }
        	            	- }
        	            	+([]v1.ExemplarQueryResult) {
        	            	 }
        	Test:       	TestDistributor/caching_unmarshal_data_disabled/series_with_exemplars
```

The test was last changed in August 2023 thought. It could be a genuine failure since it started showing up so often

**Comment by pstibrany** [comment]
I was just going to add this. I've seen this failing all morning today in 3 separate PRs.

**Comment by pracucci** [comment]
_I'm going to look into this._

**Comment by pracucci** [comment]
This reviewed again https://github.com/grafana/mimir/pull/7393 to see if I introduced a bug in `QueryExemplars()` and, just looking at the changes, I can't see it.

**Comment by pstibrany** [comment]
> _I'm going to look into this._

I'm trying to understand it via draft PR: https://github.com/grafana/mimir/pull/7454

**Comment by pracucci** [comment]
I found the cause. It's not a new thing. It's a timing issue in the ingester. The exemplar is not not ingested (but silently) because max exemplars is set to 0 in TSDB, due to a timing issue.

I'm working on a fix.

**Comment by pracucci** [comment]
Discussed different options on how to fix it with @pstibrany. Going to fix it in #7424 since it's already modifying how we compute the local exemplar limits.

**Comment by pracucci** [comment]
Fixed by https://github.com/grafana/mimir/pull/7424.
